### PR TITLE
consolidate endpoint and endpoint slice code

### PR DESCRIPTION
**Pull request by ramaraochavali**
This PR consolidates endpoint and endpoint slice code and does the following
- `updateEDS` function is unified across both so that both use common code. The varying parts have been refactored to separate interface functions
- Removed `registerEndpointsHandler` function and moved it to use the existing `registerHandler` by refactoring it to use passed in Equals func.
- The endpoint related code and tests are moved to endpoints.go from controller.go
- Renamed `endpointdiscovery.go`  to `endpointcontroller.go` to align with rest of the controller filenames in the same package.
- Handles endpoint before pod event for endpointslices (that was handled for endpoints already)

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 1]
```diff
@@ -480,21 +480,9 @@ func registerHandlers(informer cache.SharedIndexInformer, q queue.Instance, otyp
...
-		if !portsEqual(a.Subsets[i].Ports, b.Subsets[i].Ports) {
-			return false
-		}
-		if !addressesEqual(a.Subsets[i].Addresses, b.Subsets[i].Addresses) {
-			return false
-		}
-	}
-	return true
+// deepEqual is an equality func that uses costly reflect.DeepEqual to check equality.
+func deepEqual(a, b interface{}) bool {
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:484]: nit: why not pass `reflect.DeepEqual` directly?

File: pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go [hunk 2]
```diff
@@ -53,16 +53,14 @@ func (e *kubeEndpoints) Run(stopCh <-chan struct{}) {
 	e.informer.Run(stopCh)
 }
 
-// handleEvent processes the event.
-func (e *kubeEndpoints) handleEvent(name string, namespace string, event model.Event, ep interface{}, fn updateEdsFunc) error {
+// triggerConfigUpdate triggers the config update.
+func triggerConfigUpdate(c *Controller, epc kubeEndpointsController, name string, namespace string, event model.Event, ep interface{}) error {
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go:57]: nit: config update to me implies an Istio CR, not k8s registry info

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/kube/controller/controller.go]: makes sense. Changed it

File: pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go]: Changed it to `processEndpointEvent`

**Pull request merged**